### PR TITLE
new Authorization Preview API changes

### LINKED ISSUE #617: Removing token attribute from Authorizations API responses
As discussed: https://developer.github.com/changes/2014-12-08-removing-authorizations-token/

Need to be mindful of the preview phase, and make the new behaviour opt-in.

### PR DESCRIPTION
**Pull request by shiftkey**
Fixes #617

 - [x] update [List your authorizations](https://developer.github.com/v3/oauth_authorizations/#list-your-authorizations) to obsolete `token`
 - [x] update [Get a single authorization](https://developer.github.com/v3/oauth_authorizations/#get-a-single-authorization) to obsolete `token`
 - [x] switch to new endpoint if `fingerprint` provided for "Get-or-create an authorization"
 - [x] drop shared response model returned from [Get-or-create an authorization](https://developer.github.com/v3/oauth_authorizations/#get-or-create-an-authorization-for-a-specific-app) to return `token`
 - [x] update [Update an existing authorization](https://developer.github.com/v3/oauth_authorizations/#update-an-existing-authorization) to obsolete `token`
 - [x] integration tests
 - [x] address feedback

cc @joshvera

### CODE REVIEW

**Comment by shiftkey** [comment]
This is ready for review

**Comment by joshvera** [comment]
@shiftkey Sorry for missing this. Looks good to me. :+1: to merge?

**Comment by shiftkey** [comment]
@joshvera hold off until @haacked has had a chance to :eyes: over it.

I might also do an intermediate release before this - as there have been significant changes to the internals which I'd like to get out separately https://github.com/octokit/octokit.net/pull/662 https://github.com/octokit/octokit.net/pull/658

**Comment by joshvera** [comment]
:+1:

File: Octokit/Http/ApiConnection.cs [hunk 1]
```diff
@@ -251,6 +251,27 @@ public Task Put(Uri uri)
...
+        {
+            Ensure.ArgumentNotNull(uri, "uri");
+            Ensure.ArgumentNotNull(data, "data");
+
+            var response = await Connection.Put<T>(uri, data, twoFactorAuthenticationCode, accepts).ConfigureAwait(false);
+
+            return response.Body;
+        }
+
+
```
Comment (Haacked) [inline Octokit/Http/ApiConnection.cs]: Nitpick Nathan says one newline will do where two are present.

File: Octokit/Models/Request/NewAuthorization.cs [hunk 2]
```diff
@@ -11,12 +11,35 @@ namespace Octokit
...
+
+        public NewAuthorization(string note, string[] scopes)
+        {
+            Scopes = scopes;
+            Note = note;
+        }
+
+        public NewAuthorization(string note, string[] scopes, string fingerprint)
+        {
+            Scopes = scopes;
```
Comment (Haacked) [inline Octokit/Models/Request/NewAuthorization.cs]: I'm wary of passing in arrays given they can be mutated. At the very least, we should probably set `Scopes` to a copy of the arry, right?

File: Octokit/Models/Request/NewAuthorization.cs [hunk 3]
```diff
@@ -11,12 +11,35 @@ namespace Octokit
...
+            Note = note;
+            Fingerprint = fingerprint;
+        }
+
         /// <summary>
         /// Replaces the authorization scopes with this list.
         /// </summary>
         public IEnumerable<string> Scopes { get; set; }
```
Comment (Haacked) [inline Octokit/Models/Request/NewAuthorization.cs]: Are scopes always required? If so, maybe this should be readonly?

Comment (Haacked) [inline Octokit/Models/Request/NewAuthorization.cs]: Nevermind, I noticed we have an empty ctor.

File: Octokit/Models/Request/NewAuthorization.cs (hunk 2 above)
Comment (Haacked) [inline Octokit/Models/Request/NewAuthorization.cs]: I guess we have the same problem with the property setter. We can maybe worry about this separately as a general problem and not in this PR.

**Comment by Haacked** [comment]
:shipit:

File: Octokit/Models/Request/NewAuthorization.cs (hunk 3 above)
Comment (shiftkey) [inline Octokit/Models/Request/NewAuthorization.cs]: :cool:

File: Octokit/Models/Request/NewAuthorization.cs (hunk 2 above)
Comment (shiftkey) [inline Octokit/Models/Request/NewAuthorization.cs]: I'll cut a release branch for All The Breaking Changes before merging this, so we can think on this issue further.

File: Octokit/Http/ApiConnection.cs (hunk 1 above)
Comment (shiftkey) [inline Octokit/Http/ApiConnection.cs]: :ok_hand:

File: Octokit/Models/Request/NewAuthorization.cs (hunk 2 above)
Comment (shiftkey) [inline Octokit/Models/Request/NewAuthorization.cs]: This is now in-flight https://github.com/octokit/octokit.net/pull/702

**Comment by Haacked** [comment]
Is this waiting on anything or may I merge it?

**Pull request merged**